From a742cf286c77ec8ed2902f01c000417f74a35531 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Rog=C3=A9rio=20Brito?= <rbrito@ime.usp.br>
Date: Mon, 9 Dec 2019 18:12:04 -0300
Subject: [PATCH 24/27] pdfsizeopt: Consider also images that may have
 Metadata.

Usually, Metadata is stored indirectly in other objects and the tests
present so far in pdfsizeopt would test if there are any indirect references
still present after resolving some vital ones, resulting in missed images to
optimize.

See, for instance: https://github.com/pts/pdfsizeopt/issues/136

We, therefore, unset the metadata before deciding if we can/should optimize
the image in question.
---
 lib/pdfsizeopt/main.py | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/lib/pdfsizeopt/main.py b/lib/pdfsizeopt/main.py
index 1faa78a..c342eba 100644
--- a/lib/pdfsizeopt/main.py
+++ b/lib/pdfsizeopt/main.py
@@ -7426,6 +7426,15 @@ class PdfData(object):
                       r'/Device(?:RGB|Gray)[\0\t\n\r\f (<\[/])', colorspace):
         continue
 
+      # Metadata may be an indirect reference and the test after this one
+      # would make the optimizer skip files with metadata. See, for instance:
+      # https://github.com/pts/pdfsizeopt/issues/136
+      #
+      # Therefore, we unset it before testing if there are unresolved
+      # references.
+      if obj.Get('Metadata') is not None:
+        obj.Set('Metadata', None)
+
       # We've already called ResolveReferences on /Filter, /BitsPerComponent,
       # /ColorSpace, /Width, /Height, /Decode, /DecodeParms, /ImageMask.
       #
-- 
2.30.1

